### PR TITLE
provider/aws: Add aws route53 health checks

### PR DESCRIPTION
**Pull request by nicgrayson**
This PR adds a resource for route53 health checks.

### CODE REVIEW

File: builtin/providers/aws/resource_aws_route53_health_check.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+		return err
+	}
+
+	return resourceAwsRoute53HealthCheckRead(d, meta)
+}
+
+func resourceAwsRoute53HealthCheckCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).r53conn
+
+	// do we need to check if the optional fields existf before adding them?
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_health_check.go]: Typically, yes. 
For instance, if you do not check with `d.GetOk`, I _believe_ `d.Get` will give you an empty string for string values not found/required. You'll then pass that on to the API as a string pointer with value `""`, and the API will complain that isn't a valid value.

File: builtin/providers/aws/resource_aws_route53_health_check.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		CallerReference:   aws.String(time.Now().Format(time.RFC3339Nano)),
+		HealthCheckConfig: healthConfig,
+	}
+
+	wait := resource.StateChangeConf{
+		Pending:    []string{"rejected"},
+		Target:     "accepted",
+		Timeout:    5 * time.Minute,
+		MinTimeout: 1 * time.Second,
+		Refresh: func() (interface{}, string, error) {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_health_check.go]: Does this process typically take some time? If so, we're not really waiting on anything here, we just immediately reply `"accepted"` if there's no error, regardless of what the state actually is. 

If it's a quick operation that either succeeds or fails, there's no need for a refresh func. If it can take time, you should correctly populate the states here and let the function retry appropriately. At which point, you're likely wanting to call `CreateHealthCheck` separately, and maybe poll for it's status. 

Does that make sense?

**Comment by nicgrayson** [comment]
@catsby Those 2 comments should be cleared up now

File: builtin/providers/aws/resource_aws_route53_health_check.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+	if v, ok := d.GetOk("port"); ok {
+		healthConfig.Port = aws.Long(int64(v.(int)))
+	}
+
+	if v, ok := d.GetOk("resource_path"); ok {
+		healthConfig.ResourcePath = aws.String(v.(string))
+	}
+
+	wait := resource.StateChangeConf{
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_health_check.go]: This doesn't have a refresh func attached to it anymore, do this do anything?

File: builtin/providers/aws/resource_aws_route53_health_check_test.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+
+	"github.com/awslabs/aws-sdk-go/service/route53"
+)
+
+func TestAccRoute53HealthCheck(t *testing.T) {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_health_check_test.go]: We'll need to rename this to something like `TestAccRoute53HealthCheck_basic`, so that it can be ran independently from `TestAccRoute53HealthCheck_IpConfig`

**Comment by catsby** [comment]
I'm getting a panic when running either of the supplied tests, looks like some more work is needed here:

```
2015/05/18 13:40:49 [DEBUG] apply: aws_route53_health_check.foo: executing Apply
2015/05/18 13:40:49 [DEBUG] Waiting for state to become: accepted
2015/05/18 13:40:49 [TRACE] Waiting 1s before next try
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x36cf91]

goroutine 39 [running]:
github.com/hashicorp/terraform/helper/resource.func·001()
```

File: builtin/providers/aws/resource_aws_route53_record.go [hunk 5]
```diff
@@ -88,12 +88,24 @@ func resourceAwsRoute53Record() *schema.Resource {
...
-				Type:          schema.TypeSet,
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{Type: schema.TypeString},
+				Set: func(v interface{}) int {
+					return hashcode.String(v.(string))
+				},
 				ConflictsWith: []string{"alias"},
-				Elem:          &schema.Schema{Type: schema.TypeString},
 				Optional:      true,
-				Set:           schema.HashString,
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_record.go]: Is there a reason for this change? `schema.HashString` is a newer thing, shortcut for the `Set: func` you have above. Is this from a bad merge/rebase maybe?

File: builtin/providers/aws/resource_aws_route53_record.go [hunk 6]
```diff
@@ -230,10 +242,6 @@ func resourceAwsRoute53RecordRead(d *schema.ResourceData, meta interface{}) erro
 		StartRecordType: aws.String(d.Get("type").(string)),
 	}
 
-	if v, ok := d.GetOk("set_identifier"); ok {
-		lopts.StartRecordIdentifier = aws.String(v.(string))
-	}
-
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_record.go]: same question, when / why was this removed? Maybe you just need to update from `master`

File: builtin/providers/aws/resource_aws_route53_record.go (hunk 5 above)
Comment (nicgrayson) [inline builtin/providers/aws/resource_aws_route53_record.go]: It's hashstring on master: https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_route53_record.go#L96

File: builtin/providers/aws/resource_aws_route53_record.go (hunk 6 above)
Comment (nicgrayson) [inline builtin/providers/aws/resource_aws_route53_record.go]: This got renamed to SetIdentifier to match the sdk

**Comment by nicgrayson** [comment]
@catsby I've fixed up the create to remove the wait altogether. You were right, we didn't need that for this call.

File: builtin/providers/aws/resource_aws_route53_record.go (hunk 5 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_record.go]: So why are we changing it back to `func` in this pull request?

Comment (nicgrayson) [inline builtin/providers/aws/resource_aws_route53_record.go]: You are right. Fixed.

**Comment by nicgrayson** [comment]
I would love to see this merged. Let me know if anything needs to be done for that to happen.

**Comment by knuckolls** [comment]
We're using this internally on a fork of terraform. Would love to see it merged or additional feedback if needed.

**Comment by knuckolls** [comment]
Should we close this in favor of #2226? @nicgrayson @rubbish

**Comment by catsby** [comment]
#2226 seems to omit @nicgrayson from the contributors, if he's OK with that then I think we should move forward with it

**Comment by nicgrayson** [comment]
@catsby That is fine, I just want it merged :smile:

**Comment by radeksimko** [comment]
Closing in favor of #2226
